scripts/keymigrate/migrate.go

```diff
@@ -88,241 +98,411 @@ var prefixes = []struct {
+		transform: func(key keyID) (keyID, error) {
+			return convertEvidence(key, 10)
+		},
+	},
+	{
+		name:      "evidence-committed",
+		storeName: "evidence",
+		prefix:    []byte(legacyEvidenceComittedPrefix),
+		ktype:     evidenceCommittedKey,
+		transform: func(key keyID) (keyID, error) {
+			return convertEvidence(key, 9)
+		},
+	},
+	{
+		name:      "event-tx",
+		storeName: "tx_index",
+		prefix:    nil,
+		ktype:     txHeightKey,
+		transform: func(key keyID) (keyID, error) {
+			parts := bytes.Split(key, []byte("/"))
+			if len(parts) != 4 {
+				return nil, fmt.Errorf("key has %d parts rather than 4", len(parts))
+			}
+			parts = parts[1:] // drop prefix
+
+			elems := make([]interface{}, 0, len(parts)+1)
+			elems = append(elems, "tx.height")
+
+			for idx, pt := range parts {
+				val, err := strconv.Atoi(string(pt))
+				if err != nil {
+					return nil, err
+				}
+				if idx == 0 {
+					elems = append(elems, fmt.Sprintf("%d", val))
+				} else {
+					elems = append(elems, int64(val))
+				}
+			}
+
+			return orderedcode.Append(nil, elems...)
+		},
+	},
+	{
+		name:      "event-abci",
+		storeName: "tx_index",
+		prefix:    nil,
+		ktype:     abciEventKey,
+		transform: func(key keyID) (keyID, error) {
+			parts := bytes.Split(key, []byte("/"))
+
+			elems := make([]interface{}, 0, 4)
+			if len(parts) == 4 {
+				elems = append(elems, string(parts[0]), string(parts[1]))
+
+				val, err := strconv.Atoi(string(parts[2]))
+				if err != nil {
+					return nil, err
+				}
+				elems = append(elems, int64(val))
+
+				val2, err := strconv.Atoi(string(parts[3]))
+				if err != nil {
+					return nil, err
+				}
+				elems = append(elems, int64(val2))
+			} else {
+				elems = append(elems, string(parts[0]))
+				parts = parts[1:]
+
+				val, err := strconv.Atoi(string(parts[len(parts)-1]))
+				if err != nil {
+					return nil, err
+				}
+
+				val2, err := strconv.Atoi(string(parts[len(parts)-2]))
+				if err != nil {
+					return nil, err
+				}
+
+				appKey := bytes.Join(parts[:len(parts)-3], []byte("/"))
+				elems = append(elems, string(appKey), int64(val), int64(val2))
+			}
+
+			return orderedcode.Append(nil, elems...)
+		},
+	},
+	{
+		name:      "event-tx-hash",
+		storeName: "tx_index",
+		prefix:    nil,
+		ktype:     txHashKey,
+		transform: func(key keyID) (keyID, error) {
+			return orderedcode.Append(nil, "tx.hash", string(key))
+		},
+	},
+}
+
+// checkKeyType classifies a candidate key based on its structure.
+func checkKeyType(key keyID, storeName string) (keyType, error) {
+	var migrations []migrationDefinition
+	for _, m := range migrations {
+		if m.storeName != storeName {
+			continue
+		}
+		if m.prefix == nil && storeName == "tx_index" {
+			// A legacy event key has the form:
+			//
+			//    <name> / <value> / <height> / <index>
+			//
+			// Transaction hashes are stored as a raw binary hash with no prefix.
+			//
+			// Because a hash can contain any byte, it is possible (though unlikely)
```

Comment:
I believe this corner for evidence keys and hashes can no longer happen because of the table filter.